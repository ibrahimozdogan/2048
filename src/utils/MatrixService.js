class MatrixService {
    /**
     * @param {object} [config]
     * @param {Array<Array<number>>} [config.matrix]
     * @param {number} [config.newElementValue]
     * @returns {MatrixService}
     */
    constructor ({ matrix = [[]], newElementValue = 2 } = {}) {
        this.newElementValue = newElementValue;
        this._matrix = MatrixService.clone(matrix);
        this.ROTATIONS = { up: 3, right: 2, down: 1, left: 0 };

        return this;
    }

    /**
     * @returns {Array<Array<number>>}
     */
    get () {
        return this._matrix;
    }

    /**
     * @param value
     * @return {boolean}
     */
    has (value) {
        return this._matrix.filter(row => row.map(column => column === value).includes(true)).length !== 0
    }

    /**
     * @param {number} size
     * @returns {MatrixService}
     */
    create (size) {
        this._matrix = [];

        for (let i = 0; i < size; i++) {
            this._matrix[i] = [];

            for (let j = 0; j < size; j++) {
                this._matrix[i][j] = 0;
            }
        }

        return this;
    }

    /**
     * @params {number} count
     * @returns {MatrixService}
     */
    addRandomValue (count) {
        if (count === 0 || this.isMatrixFull()) {
            return this;
        }

        const size = this._matrix.length;

        let i = Math.floor(Math.random() * size);
        let j = Math.floor(Math.random() * size);

        if (this._matrix[i][j] === 0) {
            this._matrix[i][j] = this.newElementValue;

            return this.addRandomValue(count - 1);
        }

        return this.addRandomValue(count);
    }

    /**
     * @returns {boolean}
     */
    isMatrixFull () {
        return !this.has(0);
    }

    /**
     * Returns merged matrix and merged status, If matrix doesnt move, status will return false
     * @param {string} direction
     * @returns {{ matrix: Array<Array<number>>, status: boolean }}
     */
    merge (direction) {
        this.rotate(this.ROTATIONS[direction]);
        const { status, accumulatedValue } = this.shiftLeft();
        this.rotate(4 - this.ROTATIONS[direction]);

        return { matrix: this._matrix, status, accumulatedValue };
    }

    /**
     * Rotates matrix 90deg clockwise by using given times
     * @param {number} times
     */
    rotate (times) {
        let count = 0;

        while (count < times) {
            this._matrix = this._matrix.reverse();

            for (let i = 0; i < this._matrix.length; i++) {
                for (let j = 0; j < i; j++) {
                    let temp = this._matrix[i][j];

                    this._matrix[i][j] = this._matrix[j][i];
                    this._matrix[j][i] = temp;
                }
            }

            count++;
        }
    }

    /**
     * @return {{accumulatedValue: number, status: boolean}}
     */
    shiftLeft () {
        let moved = false;
        let accumulatedValue = 0;
        this._matrix = this.addProcessedProperty();

        for (let rowIndex = 0; rowIndex < this._matrix.length; rowIndex++) {
            let column = this._matrix[rowIndex];

            for (let columnIndex = 0; columnIndex < column.length - 1; columnIndex++) {
                let current = column[columnIndex];
                let next = column[columnIndex + 1];

                if (next.value === 0) {
                    continue;
                }

                if (current.value === 0) {
                    current.value = next.value;
                    next.value = 0;
                    columnIndex = -1;
                    moved = true;

                    continue;
                }

                if (current.value === next.value && !next.processed && !current.processed) {
                    current.value += next.value;
                    next.processed = true;
                    current.processed = true;
                    next.value = 0;
                    columnIndex = -1;
                    moved = true;
                    accumulatedValue += current.value;
                }
            }
        }

        this._matrix = this.removeProcessedProperty();

        return { status: moved, accumulatedValue };
    }

    /**
     * @return {Array<Array<{value: number, processed: boolean}>>}
     */
    addProcessedProperty () {
        return this._matrix.map(row => row.map(column => { return { value: column, processed: false } }));
    }

    /**
     * @return {Array<Array<number>>}
     */
    removeProcessedProperty () {
        return this._matrix.map(row => row.map(column => column.value));
    }

    /**
     * @param {object|Array} value
     * @returns {object|Array}
     */
    static clone (value) {
        return JSON.parse(JSON.stringify(value));
    }
}

export default MatrixService;
