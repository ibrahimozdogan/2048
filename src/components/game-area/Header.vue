<template>
    <div class="header">
        <div class="row">
            <div class="game-name bold">2048</div>
            <div class="score-area">
                <div class="score current-score bold">
                    <div class="title">{{translations.currentScore}}</div>
                    <div class="score">{{currentScore}}</div>
                </div>
                <div class="score best-score bold">
                    <div class="title">{{translations.bestScore}}</div>
                    <div class="score">{{bestScore}}</div>
                </div>
            </div>
        </div>
        <div class="row">
            <div>{{translations.description}}</div>
            <div class="button bold" @click="this.startGame">{{translations.newGame}}</div>
        </div>
        <div class="achievement-message" v-if="achievementCompleted">
            {{translations.achievement}}
        </div>
    </div>
</template>

<script>
    import { mapActions, mapGetters } from 'vuex';
    import CommonEnums from '../../enums/CommonEnums';
    import MatrixService from '../../utils/MatrixService';

    export default {
        computed: {
            ...mapGetters({
                currentScore: 'getCurrentScore',
                bestScore: 'getBestScore',
                blockRows: 'getBlockRows',
            }),


            achievementCompleted () {
                return new MatrixService({ matrix: this.blockRows }).has(2048);
            }
        },

        data () {
            return {
                translations: {
                    currentScore: 'Score',
                    bestScore: 'Best',
                    newGame: 'New Game',
                    description: 'Get to the 2048',
                    achievement: 'You have achieved 2048, Enjoy.'
                }
            };
        },

        methods: {
            ...mapActions([
                'setProperty'
            ]),

            startGame () {
                const blockRows = new MatrixService().create(CommonEnums.GAME_SIZE).addRandomValue(2).get();
                this.setProperty({ key: 'blockRows', value: blockRows });
            }
        }
    }
</script>

<style lang="scss" scoped>
    $huge-text: 40px;
    $big-text: 20px;
    $medium-text: 16px;
    $small-text: 12px;
    $display: flex;
    $primary-color: #FFFFFF;
    $secondary-color: #937963;
    $background-color: #937963;
    $radius: 2px;

    .header {
        width: 250px;
        margin-bottom: 30px;

        .row {
            display: $display;
            justify-content: space-between;
            align-items: center;

            .game-name {
                font-size: $huge-text;
                color: $secondary-color;
            }

            .score-area {
                display: $display;
                flex-direction: row;
                padding: 2px;
            }

            .score {
                background: $background-color;
                color: $primary-color;
                margin: 2px;
                font-size: $medium-text;
                padding: 2px 5px;
                min-width: 32px;
                border-radius: $radius;
            }

            .button {
                background: $background-color;
                color: $primary-color;
                padding: 3px 16px;
                border-radius: 2px;
                cursor: pointer;
                margin-right: 4px;
            }
        }
    }

    .bold {
        font-weight: bold;
    }

    .achievement-message {
        position: absolute;
        top: 10px;
        padding: 10px;
    }
</style>
