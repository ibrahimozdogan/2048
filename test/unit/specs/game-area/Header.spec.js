import Vuex from 'vuex';
import { mount, createLocalVue } from '@vue/test-utils'
import Header from '@/components/game-area/Header'
import MatrixService from '@/utils/MatrixService'

const localVue = createLocalVue();
localVue.use(Vuex);

describe('@components/game-area/Header.vue', () => {
    let gameStore;
    let store;
    let wrapper;
    let data;
    let currentScore = 10;
    let bestScore = 20;
    let blockRows = [[0, 2, 2048, 4], [0, 2, 0, 4], [0, 2, 0, 4], [0, 2, 0, 4]];

    beforeEach(() => {
        gameStore = {
            actions: {
                setProperty: jest.fn(),
            },
            getters: {
                getCurrentScore: () => currentScore,
                getBestScore: () => bestScore,
                getBlockRows: () => blockRows,
            },
            state: { currentScore, bestScore, blockRows },
        };
        store = new Vuex.Store({ modules: { gameStore } });
        data = {
            translations: {
                currentScore: 'Score',
                    bestScore: 'Best',
                    newGame: 'New Game',
                    description: 'Get to the 2048',
                    achievement: 'You have achieved 2048, Enjoy.'
            }
        };
        wrapper = mount(Header, {
            store,
            localVue,
            propsData: { data }
        });
    });

    describe('render', () => {
        it('renders correctly', () => {
            expect(wrapper.element).toMatchSnapshot();
        });
    });

    describe('computed', () => {
        it('should return false when 2048 is not achieved', () => {
            var hasMock = jest.spyOn(MatrixService.prototype, 'has').mockImplementation(() => false);

            expect(Header.computed.achievementCompleted.apply({ blockRows: [[]] })).toBe(false);
            expect(hasMock).toBeCalledWith(2048);

            hasMock.mockRestore();
        });

        it('should return true when 2048 is achieved', () => {
            var hasMock = jest.spyOn(MatrixService.prototype, 'has').mockImplementation(() => true);

            expect(Header.computed.achievementCompleted.apply({ blockRows: [[]] })).toBe(true);
            expect(hasMock).toBeCalledWith(2048);

            hasMock.mockRestore();
        });
    });
});
