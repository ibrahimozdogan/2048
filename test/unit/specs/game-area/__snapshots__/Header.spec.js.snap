// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@components/game-area/Header.vue render renders correctly 1`] = `
<div
  class="header"
  data="[object Object]"
>
  <div
    class="row"
  >
    <div
      class="game-name bold"
    >
      2048
    </div>
     
    <div
      class="score-area"
    >
      <div
        class="score current-score bold"
      >
        <div
          class="title"
        >
          Score
        </div>
         
        <div
          class="score"
        >
          10
        </div>
      </div>
       
      <div
        class="score best-score bold"
      >
        <div
          class="title"
        >
          Best
        </div>
         
        <div
          class="score"
        >
          20
        </div>
      </div>
    </div>
  </div>
   
  <div
    class="row"
  >
    <div>
      Get to the 2048
    </div>
     
    <div
      class="button bold"
    >
      New Game
    </div>
  </div>
   
  <div
    class="achievement-message"
  >
    
        You have achieved 2048, Enjoy.
    
  </div>
</div>
`;
